2.0/m_joinpartspam.cpp

```diff
@@ -0,0 +1,226 @@
+/*
+ * InspIRCd -- Internet Relay Chat Daemon
+ *
+ *   Copyright (C) 2016 genius3000 <genius3000@g3k.solutions>
+ *
+ * This file is a module for InspIRCd.  InspIRCd is free software: you can
+ * redistribute it and/or modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation, version 2.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "inspircd.h"
+
+/* $ModDesc: Adds channel mode +V to ban a user after x per y joins and parts/quits (join/part spam) */
+/* $ModAuthor: genius3000 */
+/* $ModAuthorMail: genius3000@g3k.solutions */
+/* $ModDepends: core 2.0 */
+
+class joinpartspamsettings
+{
+ public:
+	unsigned int cycles;
+	unsigned int secs;
+	std::map<std::string, time_t> resets;
+	std::map<std::string, unsigned int> counters;
+	time_t lastcleanup;
+
+	joinpartspamsettings(int c, int s) : cycles(c), secs(s)
+	{
+	}
+
+	/* Only actual joins are counted, but we check reset times during
+	 * both PreJoin and Join.
+	 */
+	void addcycle(const std::string& mask)
+	{
+		if (counters.count(mask) != 1)
+			resets[mask] = ServerInstance->Time() + secs;
+		else if (ServerInstance->Time() > resets[mask])
+		{
+			this->clear(mask);
+			resets[mask] = ServerInstance->Time() + secs;
+		}
+
+		++counters[mask];
+		/* Clear expired entries of non cyclers */
+		this->cleanup();
+	}
+
+	bool zapme(const std::string& mask)
+	{
+		if (counters.count(mask) != 0)
+		{
+			if (ServerInstance->Time() > resets[mask])
+				this->clear(mask);
+		}
+		return (counters[mask] >= cycles);
+	}
+
+	void cleanup()
+	{
+		/* 10 minutes should be a reasonable wait time */
+		if (ServerInstance->Time() - lastcleanup < 600)
+			return;
+
+		for (std::map<std::string, time_t>::iterator i = resets.begin(); i != resets.end(); i++)
+		{
+			if (ServerInstance->Time() > resets[i->first])
+				this->clear(i->first);
```

Comment:
This call invalidates the iterator, if the entry needs to be removed ++ the iterator before calling the function